OctopusDSC/OctopusDSCHelpers.ps1

```diff
@@ -0,0 +1,133 @@
+# Module contains shared code for OctopusDSC
+
+$octopusServerExePath = "$($env:ProgramFiles)\Octopus Deploy\Octopus\Octopus.Server.exe"
+function Get-ODSCParameter($parameters) {
+    # unfortunately $PSBoundParameters doesn't contain parameters that weren't supplied (because the default value was okay)
+    # credit to https://www.briantist.com/how-to/splatting-psboundparameters-default-values-optional-parameters/
+    $params = @{}
+    foreach ($h in $parameters.GetEnumerator()) {
+        $key = $h.Key
+        $var = Get-Variable -Name $key -ErrorAction SilentlyContinue
+        if ($null -ne $var) {
+            $val = Get-Variable -Name $key -ErrorAction Stop  | Select-Object -ExpandProperty Value -ErrorAction Stop
+            $params[$key] = $val
+        }
+    }
+    return $params
+}
+
+function Request-File {
+    [CmdletBinding()]
+    param (
+        [string]$url,
+        [string]$saveAs
+    )
+
+    $retry = $true
+    $retryCount = 0
+    $maxRetries = 5
+    $downloadFile = $true
+
+    [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12, [System.Net.SecurityProtocolType]::Tls11, [System.Net.SecurityProtocolType]::Tls
+
+    Write-Verbose "Checking to see if we have an installer at $saveas"
+    if(Test-Path $saveAs)
+    {
+        # check if we already have a matching file on disk
+        Write-Verbose "Local file exists"
+        $localHash = Get-FileHash $saveAs -Algorithm SHA256 | Select -Expand hash
+        Write-Verbose "Local SHA256 hash: $localHash"
+        $remoteHash = (Invoke-WebRequest -uri $url -Method Head -UseBasicParsing | select -expand headers).GetEnumerator()  | ? { $_.Key -eq "x-amz-meta-sha256" } | select -expand value
+        Write-Verbose "Remote SHA256 hash: $remoteHash"
+        $downloadFile = ($localHash -ne $remoteHash)
+    }
+    else
+    {
+        Write-Verbose "No local installer found"
+    }
+
+    if($downloadFile)
+    {
+        while ($retry) {
+            Write-Verbose "Downloading $url to $saveAs"
+
+            #$downloader = new-object System.Net.WebClient
+            try {
+                #$downloader.DownloadFile($url, $saveAs)
```

Comment:
Any reason for switching to `Invoke-WebRequest`? That's much slower than `WebClient`, isn't it?

Reply:
More mockable, that's about the only reason. I could have chucked the WebClient bits out into a function, I suppose. Could be something to benchmark and swap out if it seems worthwhile